File: src/lexaloud/selection.py
"""Selection capture functions.

`read_primary()` reads the X11 PRIMARY / Wayland primary selection.
`read_clipboard()` reads the CLIPBOARD / Wayland clipboard.

Neither function falls back to the other — that would be a silent footgun
(speaking a stale clipboard when the primary is empty). Callers decide which
source is meaningful in context.

Both functions apply:
- A per-subprocess timeout via `subprocess.run(..., timeout=...)` that
  terminates hung `wl-paste` / `xclip` cleanly.
- UTF-8 safe truncation to a configurable byte cap.
- An argument-list subprocess invocation (no shell=True).

Exit-code-relevant errors are raised as `SelectionError` so the CLI can map
them to exit codes 2 (empty) / 4 (oversized) / 5 (timeout or missing tool).
"""

from __future__ import annotations

import logging
import shutil
import subprocess
from dataclasses import dataclass

from .session import detect_session

log = logging.getLogger(__name__)


class SelectionError(RuntimeError):
    """Base class for capture failures."""


class SelectionEmpty(SelectionError):
    """Selection or clipboard was empty."""


class SelectionToolMissing(SelectionError):
    """Required capture tool (wl-paste/xclip) is not installed."""


class SelectionTimeout(SelectionError):
    """Capture subprocess exceeded its timeout."""


class SelectionDisplayUnavailable(SelectionError):
    """Display server is unreachable (no DISPLAY, X auth failure, etc.).

    This is distinct from `SelectionEmpty` because the CLI should tell the
    user "Lexaloud can't reach your display server" rather than the
    misleading "Select text first" notification.
    """


@dataclass
class CaptureResult:
    text: str
    truncated: bool
    original_byte_length: int
    source: str  # "primary" | "clipboard"
    tool: str  # e.g. "wl-paste --primary"


def _utf8_safe_truncate(data: bytes, max_bytes: int) -> bytes:
    """Truncate `data` to at most `max_bytes` on a UTF-8 character boundary.

    Walks back from `max_bytes` across any UTF-8 continuation bytes
    (0b10xxxxxx) so the final slice doesn't contain a partial multi-byte
    sequence. Python's slice upper bound is exclusive, so stopping at a
    lead byte cleanly excludes it. The caller additionally decodes with
    `errors="ignore"` as a belt-and-suspenders for any pre-existing
    corruption in the captured bytes.
    """
    if len(data) <= max_bytes:
        return data
    cut = max_bytes
    while cut > 0 and (data[cut] & 0xC0) == 0x80:
        cut -= 1
    return data[:cut]


# Substrings that indicate the capture tool couldn't reach the display
# server at all — as opposed to "there was no selection to read". Matched
# case-insensitively against the tool's stderr. Entries come from real
# `xclip`, `xsel`, and `wl-paste` failure messages.
_DISPLAY_FAILURE_MARKERS = (
    "can't open display",
    "cannot open display",
    "unable to open display",
    "no display",
    "display name is missing",
    "authorization",
    "not authorized",
    "wayland_display",
    "no wayland connection",
    "compositor doesn't support",
    "does not seem to support primary selection",
    "could not connect",
    "failed to connect",
)


def _run_capture(cmd: list[str], timeout_s: float) -> bytes:
    if shutil.which(cmd[0]) is None:
        raise SelectionToolMissing(f"{cmd[0]} is not installed")
    try:
        proc = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=timeout_s,
            check=False,
        )
    except subprocess.TimeoutExpired:
        raise SelectionTimeout(f"{' '.join(cmd)} timed out after {timeout_s}s")
    if proc.returncode != 0:
        # xclip/wl-paste return non-zero BOTH for "the selection is empty"
        # AND for real failures like "can't open display". Distinguish by
        # matching the stderr against known display-failure markers; real
        # empty selections produce an empty (or no) stderr. This kills the
        # "hit hotkey, get 'Select text first' — but I did select text!"
        # confusion reported in review.
        stderr = (proc.stderr or b"").decode("utf-8", errors="replace").strip()
        lowered = stderr.lower()
        if any(marker in lowered for marker in _DISPLAY_FAILURE_MARKERS):
            raise SelectionDisplayUnavailable(f"{cmd[0]} cannot reach the display server: {stderr}")
        if stderr:
            log.debug("%s exited %d with stderr: %s", cmd[0], proc.returncode, stderr)
        return b""
    return proc.stdout or b""


def _finalize(raw: bytes, max_bytes: int, source: str, tool: str) -> CaptureResult:
    original_length = len(raw)
    truncated = False
    if original_length > max_bytes:
        raw = _utf8_safe_truncate(raw, max_bytes)
        truncated = True
    text = raw.decode("utf-8", errors="ignore")
    if not text.strip():
        raise SelectionEmpty(f"{source} selection is empty")
    return CaptureResult(
        text=text,
        truncated=truncated,
        original_byte_length=original_length,
        source=source,
        tool=tool,
    )


def _pick_primary_tool(info) -> list[str]:
    """Pick the right PRIMARY capture command for the current session.

    Refuses to fall back across display servers: on Wayland without
    wl-paste, we raise `SelectionToolMissing` rather than silently using
    xclip against XWayland (which has unrelated clipboard state).
    """
    if info.is_wayland:
        if not info.wl_paste:
            raise SelectionToolMissing(
                "wl-paste is not installed. Install wl-clipboard: `sudo apt install wl-clipboard`"
            )
        return ["wl-paste", "--primary", "--no-newline"]
    if info.is_x11:
        if not info.xclip:
            raise SelectionToolMissing(
                "xclip is not installed. Install it: `sudo apt install xclip`"
            )
        return ["xclip", "-o", "-selection", "primary"]
    # session_type == "unknown"
    if info.wl_paste:
        return ["wl-paste", "--primary", "--no-newline"]
    if info.xclip:
        return ["xclip", "-o", "-selection", "primary"]
    raise SelectionToolMissing(
        "neither wl-paste nor xclip is installed; run `sudo apt install wl-clipboard xclip`"
    )


def _pick_clipboard_tool(info) -> list[str]:
    """Pick the right CLIPBOARD capture command for the current session."""
    if info.is_wayland:
        if not info.wl_paste:
            raise SelectionToolMissing(
                "wl-paste is not installed. Install wl-clipboard: `sudo apt install wl-clipboard`"
            )
        return ["wl-paste", "--no-newline"]
    if info.is_x11:
        if not info.xclip:
            raise SelectionToolMissing(
                "xclip is not installed. Install it: `sudo apt install xclip`"
            )
        return ["xclip", "-o", "-selection", "clipboard"]
    if info.wl_paste:
        return ["wl-paste", "--no-newline"]
    if info.xclip:
        return ["xclip", "-o", "-selection", "clipboard"]
    raise SelectionToolMissing(
        "neither wl-paste nor xclip is installed; run `sudo apt install wl-clipboard xclip`"
    )


def read_primary(max_bytes: int, timeout_s: float) -> CaptureResult:
    """Read the PRIMARY selection. Never touches the clipboard."""
    info = detect_session()
    cmd = _pick_primary_tool(info)
    raw = _run_capture(cmd, timeout_s)
    return _finalize(raw, max_bytes, source="primary", tool=" ".join(cmd))


def read_clipboard(max_bytes: int, timeout_s: float) -> CaptureResult:
    """Read the CLIPBOARD. Never touches the primary selection."""
    info = detect_session()
    cmd = _pick_clipboard_tool(info)
    raw = _run_capture(cmd, timeout_s)
    return _finalize(raw, max_bytes, source="clipboard", tool=" ".join(cmd))


def try_force_copy(timeout_s: float = 1.0) -> bool:
    """Best-effort synthetic Ctrl+C to copy current selection to clipboard.

    On Wayland a background `wl-paste --primary` cannot read the focused
    app's PRIMARY due to compositor security. Synthesizing Ctrl+C forces
    the app to offer the selection on CLIPBOARD, which Klipper and
    `wl-paste --no-newline` can then read. Returns True if an injector
    was invoked, False if no injector is installed.
    """
    import time

    # ydotool via uinput (daemon must be running, which it is on Plasma)
    ydotool = shutil.which("ydotool")
    if ydotool:
        try:
            subprocess.run(
                [ydotool, "key", "29:1", "46:1", "46:0", "29:0"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                timeout=min(timeout_s, 0.5),
                check=False,
            )
            time.sleep(0.20)
            return True
        except Exception as e:  # noqa: BLE001
            log.debug("ydotool force-copy failed: %s", e)

    dotool = shutil.which("dotool")
    if dotool:
        try:
            subprocess.run(
                [dotool],
                input=b"key Ctrl_L+c\n",
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                timeout=min(timeout_s, 0.5),
                check=False,
            )
            time.sleep(0.20)
            return True
        except Exception as e:  # noqa: BLE001
            log.debug("dotool force-copy failed: %s", e)

    wtype = shutil.which("wtype")
    if wtype:
        try:
            subprocess.run(
                [wtype, "-M", "ctrl", "-P", "c", "-m", "ctrl"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                timeout=min(timeout_s, 0.5),
                check=False,
            )
            time.sleep(0.20)
            return True
        except Exception as e:  # noqa: BLE001
            log.debug("wtype force-copy failed: %s", e)

    xdotool = shutil.which("xdotool")
    if xdotool:
        try:
            subprocess.run(
                [xdotool, "key", "ctrl+c"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                timeout=min(timeout_s, 0.5),
                check=False,
            )
            time.sleep(0.20)
            return True
        except Exception as e:  # noqa: BLE001
            log.debug("xdotool force-copy failed: %s", e)

    return False


def read_clipboard_via_klipper(max_bytes: int, timeout_s: float = 1.0) -> CaptureResult:
    """Read CLIPBOARD via Klipper D-Bus (privileged, works from background).

    Klipper runs as a trusted Plasma component and can read the clipboard
    even when a background `wl-paste` is denied by the compositor.
    """
    qdbus = shutil.which("qdbus6") or shutil.which("qdbus")
    if not qdbus:
        raise SelectionToolMissing("qdbus is not installed; cannot query Klipper")
    try:
        proc = subprocess.run(
            [qdbus, "org.kde.klipper", "/klipper", "org.kde.klipper.klipper.getClipboardContents"],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=timeout_s,
            check=False,
        )
    except subprocess.TimeoutExpired:
        raise SelectionTimeout(f"{qdbus} Klipper query timed out after {timeout_s}s")
    if proc.returncode != 0:
        stderr = (proc.stderr or b"").decode(errors="replace").strip()
        raise SelectionError(f"Klipper D-Bus failed: {stderr}")
    raw = (proc.stdout or b"").strip()
    # Klipper returns image placeholders like "▨ 1594x717" for images;
    # treat that as empty for TTS.
    if raw.startswith(b"\xe2\x96\xa8"):  # ▨
        raise SelectionEmpty("clipboard contains an image, not text")
    return _finalize(raw, max_bytes, source="clipboard", tool="klipper")


def try_notify(summary: str, body: str | None = None, timeout_s: float = 1.0) -> None:
    """Best-effort `notify-send`. Never raises; logs on failure."""
    notify = shutil.which("notify-send")
    if not notify:
        log.debug("notify-send not available; falling back to stderr")
        return
    # `--` terminates option parsing so a summary starting with `-` won't be
    # misinterpreted as another flag.
    args = [notify, "--app-name", "Lexaloud", "--expire-time", "3000", "--", summary]
    if body:
        args.append(body)
    try:
        subprocess.run(
            args,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            timeout=timeout_s,
            check=False,
        )
    except Exception as e:  # noqa: BLE001 — explicit best-effort swallow
        log.debug("notify-send failed: %s", e)

File: src/platform/selection.rs
use std::process::Command;
use std::time::Duration;

#[derive(Debug, thiserror::Error)]
pub enum SelectionError {
    #[error("selection is empty")]
    Empty(String),
    #[error("tool missing: {0}")]
    ToolMissing(String),
    #[error("capture timed out after {0}s")]
    Timeout(f64),
    #[error("cannot reach display server: {0}")]
    DisplayUnavailable(String),
    #[error("capture failed: {0}")]
    Other(String),
}

#[derive(Debug, Clone)]
pub struct CaptureResult {
    pub text: String,
    pub truncated: bool,
    pub original_byte_length: usize,
    pub source: String, // "primary" | "clipboard"
    pub tool: String,
}

const DISPLAY_FAILURE_MARKERS: &[&str] = &[
    "can't open display",
    "cannot open display",
    "unable to open display",
    "no display",
    "display name is missing",
    "authorization",
    "not authorized",
    "wayland_display",
    "no wayland connection",
    "compositor doesn't support",
    "does not seem to support primary selection",
    "could not connect",
    "failed to connect",
];

pub fn utf8_safe_truncate(data: &[u8], max_bytes: usize) -> Vec<u8> {
    if data.len() <= max_bytes {
        return data.to_vec();
    }
    let mut cut = max_bytes;
    while cut > 0 && (data[cut] & 0xC0) == 0x80 {
        cut -= 1;
    }
    data[..cut].to_vec()
}

fn run_capture(cmd: &[String], timeout_s: f64) -> Result<Vec<u8>, SelectionError> {
    let prog = &cmd[0];
    if which(prog).is_none() {
        return Err(SelectionError::ToolMissing(format!(
            "{} is not installed",
            prog
        )));
    }
    let mut command = Command::new(prog);
    command.args(&cmd[1..]);
    command.stdout(std::process::Stdio::piped());
    command.stderr(std::process::Stdio::piped());

    // Spawn and wait with timeout
    let mut child = command
        .spawn()
        .map_err(|e| SelectionError::Other(e.to_string()))?;
    let timeout = Duration::from_secs_f64(timeout_s);
    let start = std::time::Instant::now();
    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                let output = child
                    .wait_with_output()
                    .map_err(|e| SelectionError::Other(e.to_string()))?;
                if !status.success() {
                    let stderr = String::from_utf8_lossy(&output.stderr).to_string();
                    let lowered = stderr.to_lowercase();
                    if DISPLAY_FAILURE_MARKERS
                        .iter()
                        .any(|m| lowered.contains(&m.to_lowercase()))
                    {
                        return Err(SelectionError::DisplayUnavailable(format!(
                            "{} cannot reach the display server: {}",
                            prog,
                            stderr.trim()
                        )));
                    }
                    if !stderr.trim().is_empty() {
                        tracing::debug!(
                            "{} exited {} with stderr: {}",
                            prog,
                            status.code().unwrap_or(-1),
                            stderr.trim()
                        );
                    }
                    return Ok(Vec::new());
                }
                return Ok(output.stdout);
            }
            Ok(None) => {
                if start.elapsed() > timeout {
                    let _ = child.kill();
                    return Err(SelectionError::Timeout(timeout_s));
                }
                std::thread::sleep(Duration::from_millis(10));
            }
            Err(e) => return Err(SelectionError::Other(e.to_string())),
        }
    }
}

fn which(name: &str) -> Option<String> {
    let path = std::env::var("PATH").unwrap_or_default();
    for dir in path.split(':') {
        let p = std::path::Path::new(dir).join(name);
        if p.is_file() {
            return Some(p.to_string_lossy().to_string());
        }
    }
    None
}

fn finalize(
    raw: Vec<u8>,
    max_bytes: usize,
    source: &str,
    tool: &str,
) -> Result<CaptureResult, SelectionError> {
    let original_len = raw.len();
    let truncated = original_len > max_bytes;
    let data = if truncated {
        utf8_safe_truncate(&raw, max_bytes)
    } else {
        raw
    };
    let text = String::from_utf8_lossy(&data).to_string();
    if text.trim().is_empty() {
        return Err(SelectionError::Empty(format!(
            "{} selection is empty",
            source
        )));
    }
    Ok(CaptureResult {
        text,
        truncated,
        original_byte_length: original_len,
        source: source.to_string(),
        tool: tool.to_string(),
    })
}

#[derive(Debug, Clone)]
pub struct SessionInfo {
    pub session_type: String,
    pub desktop: String,
    pub wl_paste: Option<String>,
    pub xclip: Option<String>,
}

impl SessionInfo {
    pub fn is_wayland(&self) -> bool {
        self.session_type == "wayland"
    }
    pub fn is_x11(&self) -> bool {
        self.session_type == "x11"
    }
}

pub fn detect_session() -> SessionInfo {
    let session_type = std::env::var("XDG_SESSION_TYPE")
        .unwrap_or_else(|_| "unknown".to_string())
        .to_lowercase();
    let st = if ["wayland", "x11", "unknown"].contains(&session_type.as_str()) {
        session_type
    } else {
        "unknown".to_string()
    };
    let desktop = std::env::var("XDG_CURRENT_DESKTOP")
        .or_else(|_| std::env::var("DESKTOP_SESSION"))
        .unwrap_or_else(|_| "unknown".to_string());
    SessionInfo {
        session_type: st,
        desktop,
        wl_paste: which("wl-paste"),
        xclip: which("xclip"),
    }
}

fn pick_clipboard_tool(info: &SessionInfo) -> Result<Vec<String>, SelectionError> {
    if info.is_wayland() {
        if info.wl_paste.is_none() {
            return Err(SelectionError::ToolMissing(
                "wl-paste is not installed. Install wl-clipboard: `sudo apt install wl-clipboard`"
                    .to_string(),
            ));
        }
        return Ok(vec!["wl-paste".to_string(), "--no-newline".to_string()]);
    }
    if info.is_x11() {
        if info.xclip.is_none() {
            return Err(SelectionError::ToolMissing(
                "xclip is not installed. Install it: `sudo apt install xclip`".to_string(),
            ));
        }
        return Ok(vec![
            "xclip".to_string(),
            "-o".to_string(),
            "-selection".to_string(),
            "clipboard".to_string(),
        ]);
    }
    if info.wl_paste.is_some() {
        return Ok(vec!["wl-paste".to_string(), "--no-newline".to_string()]);
    }
    if info.xclip.is_some() {
        return Ok(vec![
            "xclip".to_string(),
            "-o".to_string(),
            "-selection".to_string(),
            "clipboard".to_string(),
        ]);
    }
    Err(SelectionError::ToolMissing(
        "neither wl-paste nor xclip is installed; run `sudo apt install wl-clipboard xclip`"
            .to_string(),
    ))
}

pub fn read_clipboard(max_bytes: usize, timeout_s: f64) -> Result<CaptureResult, SelectionError> {
    let info = detect_session();
    let cmd = pick_clipboard_tool(&info)?;
    let raw = run_capture(&cmd, timeout_s)?;
    finalize(raw, max_bytes, "clipboard", &cmd.join(" "))
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn utf8_truncate_basic() {
        let data = "hello world".as_bytes();
        let t = utf8_safe_truncate(data, 5);
        assert_eq!(String::from_utf8_lossy(&t), "hello");
    }
    #[test]
    fn utf8_truncate_multibyte() {
        // "héllo" -> h(1) é(2) l(1) l(1) o(1) = 6 bytes; cut at 3 lands
        // exactly after é, so the result must be "hé", not "h".
        let data = "héllo".as_bytes();
        let t = utf8_safe_truncate(data, 3);
        assert_eq!(String::from_utf8(t).unwrap(), "hé");
        // Cut at 2 would split é; the truncate must back off to "h".
        let t2 = utf8_safe_truncate(data, 2);
        assert_eq!(String::from_utf8(t2).unwrap(), "h");
    }
    #[test]
    fn finalize_truncated() {
        let raw = b"hello world".to_vec();
        let res = finalize(raw, 5, "primary", "test").unwrap();
        assert!(res.truncated);
        assert_eq!(res.original_byte_length, 11);
        assert_eq!(res.text, "hello");
    }
    #[test]
    fn finalize_empty_fails() {
        let raw = b"   \n".to_vec();
        let res = finalize(raw, 100, "primary", "test");
        assert!(res.is_err());
    }
}
